agent/src/apiclient.rs

```diff
@@ -0,0 +1,296 @@
+
+    #[snafu(display("Failed to deserialize update status: '{}'", source))]
+    UpdateStatusContent { source: serde_json::Error },
+
+    #[snafu(display("Failed to deserialize Os info: '{}'", source))]
+    GetMostRecentCommand { source: serde_json::Error },
+
+    #[snafu(display("failed to refresh updates or update action performed out of band"))]
+    RefreshUpdate {},
+
+    #[snafu(display("failed to prepare update or update action performed out of band"))]
+    PrepareUpdate {},
+
+    #[snafu(display("failed to activate update or update action performed out of band"))]
+    Update {},
+
+    #[snafu(display(
+        "unexpected update state: '{:?}', expecting state to be '{}'. update action performed out of band?",
+         update_state, expect_state
+    ))]
+    UpdateStage {
+        expect_state: String,
+        update_state: UpdateState,
+    },
+}
+
+// TODO: docstrings on those enum
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+pub enum UpdateState {
+    Idle,
+    Available,
+    Staged,
+    Ready,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum UpdateCommand {
+    Refresh,
+    Prepare,
+    Activate,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+enum CommandStatus {
+    Success,
+    Failed,
+    Unknown,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct StagedImage {
+    image: Option<UpdateImage>,
+    next_to_boot: bool,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct CommandResult {
+    cmd_type: UpdateCommand,
+    cmd_status: CommandStatus,
+    timestamp: String,
+    exit_status: u32,
+    stderr: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct UpdateImage {
+    arch: String,
+    version: String,
+    variant: String,
+}
+
+pub struct OsInfo {
+    pub version_id: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct UpdateStatus {
+    update_state: UpdateState,
+    available_updates: Vec<String>,
+    chosen_update: Option<UpdateImage>,
+    active_partition: Option<StagedImage>,
+    staging_partition: Option<StagedImage>,
+    most_recent_command: CommandResult,
+}
+
+pub fn get_raw_args(mut args: Vec<String>) -> Vec<String> {
+    let mut subcommand_args = Vec::new();
+    subcommand_args.push("raw".to_string());
+    subcommand_args.push("-u".to_string());
+    subcommand_args.append(&mut args);
+
+    return subcommand_args;
+}
+
+/// TODO
+/// a helper function to invoke apiclient command
+/// Nit: I'd add some details to this docstring about the fact that the API client is volume mounted into the agent container.
+pub fn invoke_apiclient(args: Vec<String>) -> Result<Output> {
+    let output = Command::new("/bin/apiclient")
+        .args(&args)
+        .output()
+        .context(ApiClientCommand { args })?;
+
+    Ok(output)
+}
+
+pub fn get_update_status() -> Result<UpdateStatus> {
+    let mut raw_args = Vec::new();
+    raw_args.push("/updates/status".to_string());
```

Comment:
Nit: `"/updates/status"`, should be a constant, and then:

```suggestion
    let mut raw_args = vec![UPDATES_STATUS_URI.to_string()];
```